crates/reporting/src/error/type.rs

```diff
@@ -3126,14 +3155,220 @@ fn diff_record<'b>(
     }
 }
 
+/// This is a helper for should_show_field_diff - see its doc comment for details.
+fn should_show_diff(t1: &ErrorType, t2: &ErrorType) -> bool {
+    use ErrorType::*;
+
+    match (t1, t2) {
+        (Type(sym1, types1), Type(sym2, types2)) => {
+            if sym1 != sym2 || types1.len() != types2.len() {
+                return true;
+            }
+
+            types1
+                .iter()
+                .zip(types2.iter())
+                .any(|(t1, t2)| should_show_diff(t1, t2))
+        }
+        (Infinite, Infinite) | (Error, Error) => false,
+        (FlexVar(v1), FlexVar(v2)) => v1 != v2,
+        (RigidVar(v1), RigidVar(v2)) => v1 != v2,
+        (FlexAbleVar(v1, set1), FlexAbleVar(v2, set2)) => {
+            if v1 != v2 || set1.len() != set2.len() {
+                return true;
+            }
+
+            set1.sorted_iter()
+                .zip(set2.sorted_iter())
+                .any(|(t1, t2)| t1 != t2)
+        }
+        (RigidAbleVar(v1, set1), RigidAbleVar(v2, set2)) => {
+            if v1 != v2 || set1.len() != set2.len() {
+                return true;
+            }
+
+            set1.sorted_iter()
+                .zip(set2.sorted_iter())
+                .any(|(t1, t2)| t1 != t2)
+        }
+        (Record(fields1, ext1), Record(fields2, ext2)) => {
+            if fields1.len() != fields1.len() || ext1 != ext2 {
+                return true;
+            }
+
+            fields1
+                .iter()
+                .zip(fields2.iter())
+                .any(|((name1, f1), (name2, f2))| name1 != name2 || should_show_field_diff(f1, f2))
+        }
+        (TagUnion(tags1, ext1, _polarity1), TagUnion(tags2, ext2, _polarity2)) => {
+            // If two tag unions differ only in polarity, don't show that as a diff;
+            // polarity is invisible to the reader!
```

Comment:
Did you run into a case where the polarity does differ? The polarity should always be the same between two tag unions we're looking at to compare - if not that's a bug! (may be worth a debug_assert)